src/storage/txn/flow_controller.rs

```diff
@@ -0,0 +1,1072 @@
+    // decided based on the statistics of the throttle CF. If the multiple CFs
+    // exceed the threshold, choose the larger one.
+    throttle_cf: Option<String>,
+    // The target flow of L0, the algorithm's goal is to make flush flow close
+    // to L0 target flow.
+    l0_target_flow: f64,
+    // The number of L0 files when the last update of L0 target flow.
+    last_target_file: Option<u64>,
+    // Discard ratio is decided by pending compaction bytes, it's the ratio to
+    // drop write requests(return ServerIsBusy to TiDB) randomly.
+    discard_ratio: Arc<AtomicU64>,
+
+    engine: E,
+    limiter: Arc<Limiter>,
+    // Records the foreground write flow at scheduler level of last few seconds.
+    write_flow_recorder: Smoother<30>,
+    last_record_time: Instant,
+}
+
+impl<E: KvEngine> FlowChecker<E> {
+    pub fn new(
+        config: &FlowControlConfig,
+        engine: E,
+        discard_ratio: Arc<AtomicU64>,
+        limiter: Arc<Limiter>,
+    ) -> Self {
+        let mut cf_checkers = map![];
+
+        for cf in engine.cf_names() {
+            cf_checkers.insert(cf.to_owned(), CFFlowChecker::default());
+        }
+
+        Self {
+            soft_pending_compaction_bytes_limit: config.soft_pending_compaction_bytes_limit,
+            hard_pending_compaction_bytes_limit: config.hard_pending_compaction_bytes_limit,
+            memtables_threshold: config.memtables_threshold,
+            l0_files_threshold: config.l0_files_threshold,
+            engine,
+            discard_ratio,
+            limiter,
+            write_flow_recorder: Smoother::default(),
+            cf_checkers,
+            throttle_cf: None,
+            l0_target_flow: 0.0,
+            last_target_file: None,
+            last_record_time: Instant::now_coarse(),
+        }
+    }
+
+    fn start(self, rx: Receiver<Msg>, flow_info_receiver: Receiver<FlowInfo>) -> JoinHandle<()> {
+        Builder::new()
+            .name(thd_name!("flow-checker"))
+            .spawn(move || {
+                tikv_alloc::add_thread_memory_accessor();
+                let mut checker = self;
+                let mut deadline = std::time::Instant::now();
+                let mut spare_ticks = 0;
+                let mut enabled = true;
+                loop {
+                    match rx.try_recv() {
+                        Ok(Msg::Close) => break,
+                        Ok(Msg::Disable) => {
+                            enabled = false;
+                            checker.reset_statistics();
+                        }
+                        Ok(Msg::Enable) => {
+                            enabled = true;
+                        }
+                        Err(_) => {}
+                    }
+
+                    if !enabled {
+                        // do nothing, just consume the flow info channel
+                        let _ = flow_info_receiver.recv();
+                        continue;
+                    }
```

Comment:
Since we are still receiving flow_info, would it be better to do the check as well? No much overhead I presume.

Reply:
do the check, then what?